aggregator/options.go

```diff
@@ -95,52 +94,63 @@ var (
 	defaultGaugeSuffixOverride = map[policy.AggregationType][]byte{
 		policy.Last: nil,
 	}
+
+	// By default writes are buffered for 10 minutes before traffic is cut over to a shard
+	// in case there are issues with a new instance taking over shards.
+	defaultBufferDurationBeforeShardCutover = 10 * time.Minute
+
+	// By default writes are buffered for one hour after traffic is cut off in case there
+	// are issues with the instances taking over the shards and as such we need to switch
+	// the traffic back to the previous owner of the shards immediately.
+	defaultBufferDurationAfterShardCutoff = time.Hour
 )
 
 type options struct {
 	// Base options.
-	defaultCounterAggregationTypes policy.AggregationTypes
-	defaultTimerAggregationTypes   policy.AggregationTypes
-	defaultGaugeAggregationTypes   policy.AggregationTypes
-	metricPrefix                   []byte
-	counterPrefix                  []byte
-	timerPrefix                    []byte
-	gaugePrefix                    []byte
-	sumSuffix                      []byte
-	sumSqSuffix                    []byte
-	meanSuffix                     []byte
-	lastSuffix                     []byte
-	minSuffix                      []byte
-	maxSuffix                      []byte
-	countSuffix                    []byte
-	stdevSuffix                    []byte
-	medianSuffix                   []byte
-	timerQuantileSuffixFn          QuantileSuffixFn
-	timeLock                       *sync.RWMutex
-	clockOpts                      clock.Options
-	instrumentOpts                 instrument.Options
-	streamOpts                     cm.Options
-	placementWatcherOpts           services.StagedPlacementWatcherOptions
-	instanceID                     string
-	shardFn                        ShardFn
-	flushManager                   FlushManager
-	minFlushInterval               time.Duration
-	maxFlushSize                   int
-	flushHandler                   Handler
-	entryTTL                       time.Duration
-	entryCheckInterval             time.Duration
-	entryCheckBatchPercent         float64
-	maxTimerBatchSizePerWrite      int
-	defaultPolicies                []policy.Policy
-	electionManager                ElectionManager
-	resignTimeout                  time.Duration
-	entryPool                      EntryPool
-	counterElemPool                CounterElemPool
-	timerElemPool                  TimerElemPool
-	gaugeElemPool                  GaugeElemPool
-	bufferedEncoderPool            msgpack.BufferedEncoderPool
-	aggTypesPool                   policy.AggregationTypesPool
-	quantilesPool                  pool.FloatsPool
+	defaultCounterAggregationTypes   policy.AggregationTypes
```

Comment:
super nit: this struct is pretty large, does it make sense to group some of the fields into their own smaller options and include them, kind of like we do with `clockOpts`, `intstrumentOpts`, and others? For example, it looks like the prefixes and suffixes could be grouped together in their own options type.

Reply:
Possibly, although there are derived fields for performance reasons that are computed from the base fields and are updated when the base fields are set. Those derived fields will need to be grouped together w/ the base fields for that to continue to work. I can look at it in a separate refactoring PR.